File: core/src/main/java/de/qabel/core/drop/DropMessage.java
package de.qabel.core.drop;

import de.qabel.core.config.Entity;
import de.qabel.core.config.Identity;

import java.io.Serializable;
import java.util.Date;
import java.util.UUID;

public class DropMessage implements Serializable {
    /**
     * Acknowledge ID indicating that the sender does not
     * wish to receive an acknowledgement.
     */
    public static final String NOACK = "0";
    /**
     * Model object reserved for internal drop protocol purposes only.
     */
    public static final String INTERNAL_MODEL_OBJECT = "drop";

    private static final int VERSION = 1;


    private Date created;
    private String acknowledgeId;
    private Entity sender;
    private String senderKeyId;
    private String dropPayload;
    private String dropPayloadType;

    private DropMessageMetadata dropMessageMetadata;

    public DropMessage(Entity sender, String dropPayload, String dropPayloadType) {
        this.sender = sender;
        this.senderKeyId = sender.getKeyIdentifier();
        this.dropPayload = dropPayload;
        this.dropPayloadType = dropPayloadType;
        created = new Date();
        acknowledgeId = NOACK;
    }

    /**
     * Constructor used for deserialization.
     * registerSender has to be called to complete creation.
     */
    DropMessage(String senderKeyId, String dropPayload, String dropPayloadType, Date created, String acknowledgeId, DropMessageMetadata dropMessageMetadata) {
        this.senderKeyId = senderKeyId;
        this.dropPayload = dropPayload;
        this.dropPayloadType = dropPayloadType;
        this.created = created;
        this.acknowledgeId = acknowledgeId;
        this.dropMessageMetadata = dropMessageMetadata;
    }

    public static int getVersion() {
        return VERSION;
    }

    public Date getCreationDate() {
        return created;
    }

    public String getAcknowledgeID() {
        return acknowledgeId;
    }

    public Entity getSender() {
        return sender;
    }

    public String getSenderKeyId() {
        return senderKeyId;
    }

    /**
     * Register the given Entity as sender of this drop message
     * and check if it matches the senderKeyId.
     * This is used to complete the deserialization of DropMessage.
     *
     * @param sender Entity to be registered as sender.
     * @return true if the given sender matches the senderKeyId, otherwise false.
     */
    public boolean registerSender(Entity sender) {
        if (!senderKeyId.equals(sender.getKeyIdentifier())) {
            return false;
        }
        this.sender = sender;
        return true;
    }

    public String getDropPayload() {
        return dropPayload;
    }

    public String getDropPayloadType() {
        return dropPayloadType;
    }

    /**
     * Enable/disable acknowledge request for this drop message.
     * Acknowledging is disabled by default.
     *
     * @param enabled true enables acknowledging
     */
    public void enableAcknowledging(boolean enabled) {
        if (enabled && acknowledgeId.equals(NOACK)) {
            acknowledgeId = generateAcknowledgeId();
        } else if (!enabled) {
            acknowledgeId = NOACK;
        }
    }

    private static String generateAcknowledgeId() {
        return UUID.randomUUID().toString();
    }

    public DropMessageMetadata getDropMessageMetadata() {
        return dropMessageMetadata;
    }

    public void setDropMessageMetadata(DropMessageMetadata dropMessageMetadata){
        this.dropMessageMetadata = dropMessageMetadata;
    }
}

File: core/src/main/java/de/qabel/core/drop/DropMessage.kt
package de.qabel.core.drop

import de.qabel.core.config.Entity
import de.qabel.core.config.Identity

import java.io.Serializable
import java.util.Date
import java.util.UUID

class DropMessage : Serializable {


    var creationDate: Date
        private set
    var acknowledgeID: String
        private set
    var sender: Entity? = null
        private set
    var senderKeyId: String
        private set
    var dropPayload: String
        private set
    var dropPayloadType: String
        private set

    var dropMessageMetadata: DropMessageMetadata? = null

    constructor(sender: Entity, dropPayload: String, dropPayloadType: String) {
        this.sender = sender
        this.senderKeyId = sender.keyIdentifier
        this.dropPayload = dropPayload
        this.dropPayloadType = dropPayloadType
        creationDate = Date()
        acknowledgeID = NOACK
    }

    /**
     * Constructor used for deserialization.
     * registerSender has to be called to complete creation.
     */
    internal constructor(senderKeyId: String, dropPayload: String, dropPayloadType: String, created: Date,
                         acknowledgeId: String, dropMessageMetadata: DropMessageMetadata?) {
        this.senderKeyId = senderKeyId
        this.dropPayload = dropPayload
        this.dropPayloadType = dropPayloadType
        this.creationDate = created
        this.acknowledgeID = acknowledgeId
        this.dropMessageMetadata = dropMessageMetadata
    }

    /**
     * Register the given Entity as sender of this drop message
     * and check if it matches the senderKeyId.
     * This is used to complete the deserialization of DropMessage.

     * @param sender Entity to be registered as sender.
     * *
     * @return true if the given sender matches the senderKeyId, otherwise false.
     */
    fun registerSender(sender: Entity): Boolean {
        if (senderKeyId != sender.keyIdentifier) {
            return false
        }
        this.sender = sender
        return true
    }

    /**
     * Enable/disable acknowledge request for this drop message.
     * Acknowledging is disabled by default.

     * @param enabled true enables acknowledging
     */
    fun enableAcknowledging(enabled: Boolean) {
        if (enabled && acknowledgeID == NOACK) {
            acknowledgeID = generateAcknowledgeId()
        } else if (!enabled) {
            acknowledgeID = NOACK
        }
    }

    companion object {
        /**
         * Acknowledge ID indicating that the sender does not
         * wish to receive an acknowledgement.
         */
        val NOACK = "0"
        /**
         * Model object reserved for internal drop protocol purposes only.
         */
        val INTERNAL_MODEL_OBJECT = "drop"

        val version = 1

        private fun generateAcknowledgeId(): String {
            return UUID.randomUUID().toString()
        }
    }
}
